Exploratory prototype for pre-authZN

groldan opened the pull request:
New module `georchestra-authn-authz` that will take care of
receiving all the necessary user information (properties and
credentials) from security-proxy and keep the GeoNetwork users in-sync
without cron jobs.

> Works against https://github.com/georchestra/georchestra/pull/3476
https://github.com/georchestra/georchestra/tree/georchestra-gn4-4.0.5

---

# GeoNetwork integration with Georchestra pre-authentication and authorization

This module integrates *GeoNetwork* user details and authorization into *geOrchestra*,
with *geOrchestra* providing all the necessary pre-authentication and authorization
credentials on each HTTP request, through `security-proxy` application gateway's
HTTP request headers.

That means there's no need to configure other GeoNetwork components to match or
synchronize the User and Organization properties and credentials, with
geOrchestra acting as the one source of truth for user management.

## Background

*GeoNetwork* **requires** its users to be defined in the same PostgreSQL database
where it stores the metadata records and other stuff.

*geOrchestra* maintains its own user database in LDAP, and is to be taken as the
canonical representation of user details.

Traditionally, *GeoNetwork* is customized to fetch user details from LDAP, and
periodically synchronize the users in Postgres from LDAP. This comes with some
caveats, as the synchronization process is performed in a `cron-like` fashion,
and hence there's a time window where Postgres users are potentially out of sync
with LDAP users.

Also, *geOrchestra* users are identified by their username, and GeoNetwork users
by a unique integer id. Since a username may be changed, the synchronization process
will leave Postgres users dangling and create new ones?

Finally, the proliferation of direct access to LDAP to get user details
among the different *geOrchestra* applications only makes the whole echosystem
harder to maintain, whereas ideally it should be an implementation detail managed
by a single component/application (`console`?), and have a system boundary
for the rest of applications (interface and common object model, for example).

Within geOrchestra's environment, all HTTP requests to GeoNetwork come from the 
`security-proxy` application gateway, which now has the ability to send an 
arbitrary (configurable) set of user properties as HTTP headers. This opens up
the possibility for backend services to be completely decoupled from LDAP and
be given all the information they need about users/groups.

The component that integrates geOrchestra authZN with GeoNetwork can automate
the link between GN User entities and pre-authenticated users through a provided
unique identifier. 

The link from "geOrchestra user id" to "GeoNetwork user id" can be managed as a
JPA entity, and the current-ness of the Postgres user entity with regard to the
provided user details be quickly checked on-demand (e.g. through a "last-updated"
property, or heuristically by comparing relevant user properties), hence eliminating
the need for cron-like sychronization processes and the attached out-of-sync time
window.

## Risks

- `security-proxy` **must** be able to provide a unique user identifier,
`sec-username` is not good engouh as it can be changed. We're not sure yet
whether a stable UUID will be provided by LDAP? Mitigation strategy: by getting
rid of LDAP in GeoNetwork and using only the provided pre-auth credentials, we can
do whatever we want in geOrchestra's `console` app to have stable user ids, without
affecting other systems.
- The component integrates geOrchestra authZN with GeoNetwork can automate
the link

## Architecture

With `org.georchestra.geonetwork.security` as base package:

````
authentication
     |
     | <uses>
     V
 integration
     |
     | <uses>
     V
 repository
````


## Pre-authentication

- `security-proxy` sends the necessary HTTP request headers (`sec-*`) to
handle the pre-authenticated user's identity and privileges.

- GeoNetwork's `security-proxy` headsers **MUST** be configured to send
a user's unique identifier (`sec-userid:string`) among the authN/authZ headers, 
which is used to map GeoNetwork users in its PostgreSQL database 1:1 
with geOrchestra users.

- The `security-proxy-spring-integration` dependency provides the 
pre-authentication Servlet filter.

---

TODO:

* Remove all ldap related configuration. From `web/src/main/webapp/WEB-INF/config-security/config-security-georchestra.xml`:
    * `ldapAuthenticationProviderPostProcessor`
    * `ldapUtils`
    * `contextSource`
    * `ldapAuthProvider`
    * `ldapUserContextMapper`
    * `ldapManager`/`ldapUserDetailsService`
    * `ldapSynchronizer`
    * `userMapper`
    * `ldapUserSearch`
* Configure georchestra's pre-auth filter and on-demand user synchronizer:
  * From `web/src/main/webapp/WEB-INF/config-security/config-security-georchestra.xml`, replace `preAuthenticationFilter` and `preFilterAuthenticationProvider` by the ones that take full AuthN/AuthZ credentials from request headers and forces matching the geonetwork user info with the georchestra user credentials.
  
* Replace all ldap related configuration in geonetwork's data directory (from georchestra), it would only require sending the
appropriate sec-* headers (like with the datafeeder app).

groldan committed:
Exploratory prototype for pre-authZN

New module `georchestra-authn-authz` that will take care of
receiving all the necessary user information (properties and
credentials) from security-proxy and keep the GeoNetwork users in-sync
without cron jobs.

fvanderbiest commented:
Thanks for the proposal Gabriel !

I like **a lot** the fact that we decouple GeoNetwork from the LDAP !
One thing though. GeoNetwork groups (currently synced with ldap orgs by default / or with a selection of ldap roles in legacy setups) are required for the admin to manage metadata ACLs, even before a single user belonging to this specific Org or having this exact Role connects onto GeoNetwork. How will it be covered ?
Not sure we can fully decouple things...

groldan commented:
Hey @fvanderbiest, just to make sure I get it right. That sync'ing of GN groups with geOr roles is performed by the same ldap sync cron job, right? or maybe when it first starts up, but the same synchronizer?

I think the solution would still be (as suggested, maybe not too clearly) to actually have a "system boundary" (which would be incarnated by an interface, a service (in `console` app maybe?), and a domain model (think openapi generated).

From that point on, there could be different integration strategies. We're maybe not fully set up to do event streaming, yet geonetwork could just request the roles through the API at startup and get the initial sync if need be, and/or be called in proactivelly by console upon role creation/change/deletion. 

Just thinking out loud... but still think decoupling from LDAP is a worthwhile goal.

fvanderbiest commented:
> That sync'ing of GN groups with geOr roles is performed by the same ldap sync cron job, right?

Yeah, it happens in the same cron job

> We're maybe not fully set up to do event streaming

Nope, but this might be something to consider, because fetching the roles at startup will not match platform admin expectations

ghost commented:
- We do not only have the LDAP integrated within geOrchestra. We also have projects where the LDAP is upstream and not in our hands. But I think we can assume that USER-IDs are stable there. For example, we could copy username to userid in these cases within the security-proxy.

- Group-Sync: Groups must be synchronized independently of user-logins: Even if no user at all logs into GeoNetwork, all groups must get synchronised, such that a GeoNetwork administrator can use these groups for authorization.

- If we keep these two points in mind, I think it is good and possible to have an abstraction that factors out the user/group-synchronization.

groldan commented:
Björn Höfling comments:

- We do not only have the LDAP integrated within geOrchestra. We also have projects where the LDAP is upstream and not in our hands. But I think we can assume that USER-IDs are stable there. For example, we could copy username to userid in these cases within the security-proxy.

- Group-Sync: Groups must be synchronized independently of user-logins: Even if no user at all logs into GeoNetwork, all groups must get synchronised, such that a GeoNetwork administrator can use these groups for authorization.

- If we keep these two points in mind, I think it is good and possible to have an abstraction that factors out the user/group-synchronization.

groldan committed:
WIP integrating with GeoNetwork, having issues with JPA

- security-proxy sends all the pre-authZN info, see docker/config/security-proxy/headers-mapping.properties
- GeorchestraPreAuthenticationFilter takes GeorchestraUserDetails and
  matches a GN User (creating or updating it as necessary)
- All the LDAP configuration is removed, Georchestra integration shall
  work against a higher level abstraction. Initial groups
synchronization will be done against an interface rather than directly
against LDAP, whose implementation should be a REST API (probably in
console)

groldan committed:
REVISIT: Move JPAUserLink to gn-domain

Couldn't make JPA load the JPAUserLink entity from a separate .jar.
Ideally this entity should be loaded at run/startup time and contributed
to the entity manager.

groldan committed:
tmp

groldan commented:
Closing, superseded by #182 due to change in target branch
